File: rxbinding-design/src/main/java/com/jakewharton/rxbinding2/support/design/widget/SnackbarDismissesObservable.java
package com.jakewharton.rxbinding2.support.design.widget;

import android.support.design.widget.Snackbar;
import android.support.design.widget.Snackbar.Callback;
import io.reactivex.Observable;
import io.reactivex.Observer;
import io.reactivex.android.MainThreadDisposable;

import static com.jakewharton.rxbinding2.internal.Preconditions.checkMainThread;

final class SnackbarDismissesObservable extends Observable<Integer> {
  private final Snackbar view;

  SnackbarDismissesObservable(Snackbar view) {
    this.view = view;
  }

  @Override protected void subscribeActual(Observer<? super Integer> observer) {
    if (!checkMainThread(observer)) {
      return;
    }
    Listener listener = new Listener(view, observer);
    observer.onSubscribe(listener);
    view.setCallback(listener.callback);
  }

  final class Listener extends MainThreadDisposable {
    private final Snackbar snackbar;
    private final Callback callback;

    Listener(Snackbar snackbar, final Observer<? super Integer> observer) {
      this.snackbar = snackbar;
      this.callback = new Callback() {
        @Override public void onDismissed(Snackbar snackbar, int event) {
          if (!isDisposed()) {
            observer.onNext(event);
          }
        }
      };
    }

    @Override protected void onDispose() {
      snackbar.setCallback(null);
    }
  }
}

File: rxbinding-material/src/main/java/com/jakewharton/rxbinding3/material/SnackbarDismissesObservable.kt
@file:JvmName("RxSnackbar")
@file:JvmMultifileClass

package com.jakewharton.rxbinding3.material

import androidx.annotation.CheckResult
import com.google.android.material.snackbar.Snackbar
import com.google.android.material.snackbar.Snackbar.Callback
import com.jakewharton.rxbinding2.internal.Preconditions.checkMainThread
import io.reactivex.Observable
import io.reactivex.Observer
import io.reactivex.android.MainThreadDisposable

/**
 * Create an observable which emits the dismiss events from `view`.
 *
 * *Warning:* The created observable keeps a strong reference to `view`. Unsubscribe
 * to free this reference.
 */
@CheckResult
fun Snackbar.dismisses(): Observable<Int> {
  return SnackbarDismissesObservable(this)
}

private class SnackbarDismissesObservable(
  private val view: Snackbar
) : Observable<Int>() {

  override fun subscribeActual(observer: Observer<in Int>) {
    if (!checkMainThread(observer)) {
      return
    }
    val listener = Listener(view, observer)
    observer.onSubscribe(listener)
    view.addCallback(listener.callback)
  }

  private class Listener(
    private val snackbar: Snackbar,
    observer: Observer<in Int>
  ) : MainThreadDisposable() {
    val callback = object : Callback() {
      override fun onDismissed(snackbar: Snackbar?, event: Int) {
        if (!isDisposed) {
          observer.onNext(event)
        }
      }
    }

    override fun onDispose() {
      snackbar.removeCallback(callback)
    }
  }
}
